src/cascadia/TerminalControl/TSFInputControl.cpp

```diff
@@ -0,0 +1,379 @@
+#include "pch.h"
+#include "TSFInputControl.h"
+#include "TSFInputControl.g.cpp"
+
+using namespace winrt::Windows::Foundation;
+using namespace winrt::Windows::Graphics::Display;
+using namespace winrt::Windows::UI::Core;
+using namespace winrt::Windows::UI::Text;
+using namespace winrt::Windows::UI::Text::Core;
+using namespace winrt::Windows::UI::Xaml;
+
+// hack to account for offset caused by tabs
+#define TABS_OFFSET 38
+
+namespace winrt::Microsoft::Terminal::TerminalControl::implementation
+{
+    TSFInputControl::TSFInputControl() :
+        _editContext{ nullptr }
+    {
+        _Create();
+    }
+
+    // Method Description:
+    // - Creates XAML controls for displaying user input and hooks up CoreTextEditContext handlers
+    //   for handling text input from the Text Services Framework.
+    // Arguments:
+    // - <none>
+    // Return Value:
+    // - <none>
+    void TSFInputControl::_Create()
+    {
+        // TextBlock for user input form TSF
+        _textBlock = Controls::TextBlock();
+        _textBlock.Visibility(Visibility::Collapsed);
+        _textBlock.IsTextSelectionEnabled(false);
+        _textBlock.TextDecorations(TextDecorations::Underline);
+
+        // Canvas for controlling exact position of the TextBlock
+        _canvas = Windows::UI::Xaml::Controls::Canvas();
+        _canvas.Visibility(Visibility::Collapsed);
+
+        // add the Textblock to the Canvas
+        _canvas.Children().Append(_textBlock);
+
+        // set the content of this control to be the Canvas
+        this->Content(_canvas);
+
+        // Create a CoreTextEditingContext for since we are acting like a custom edit control
+        auto manager = Core::CoreTextServicesManager::GetForCurrentView();
+        _editContext = manager.CreateEditContext();
+
+        // sets the Input Pane display policy to Manual for now so that it can manually show the
+        // software keyboard when the control gains focus and dismiss it when the control loses focus.
+        // TODO investigate if we can set this to Automatic

```

Comment:
leave an issue ID with a TODO

Reply:
I've opened #3639 to track the investigation if it's needed or not.  Code has issue in it as well.